Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift

```diff
@@ -0,0 +1,152 @@
+    private let productsManager: ProductsManager
+
+    init(receiptFetcher: ReceiptFetcher,
+         introEligibilityCalculator: IntroEligibilityCalculator,
+         backend: Backend,
+         identityManager: IdentityManager,
+         operationDispatcher: OperationDispatcher,
+         productsManager: ProductsManager) {
+        self.receiptFetcher = receiptFetcher
+        self.introEligibilityCalculator = introEligibilityCalculator
+        self.backend = backend
+        self.identityManager = identityManager
+        self.operationDispatcher = operationDispatcher
+        self.productsManager = productsManager
+    }
+
+    // swiftlint:disable line_length
+    func checkTrialOrIntroPriceEligibilityFromOptimalStoreKitVersion(_ productIdentifiers: [String],
+                                                                     completionBlock receiveEligibility: @escaping ReceiveIntroEligibilityBlock) {
+        // swiftlint:enable line_length
+        if #available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *) {
+            sk2CheckTrialOrIntroPriceEligibility(productIdentifiers, completionBlock: receiveEligibility)
+        } else {
+            sk1CheckTrialOrIntroPriceEligibility(productIdentifiers, completionBlock: receiveEligibility)
+        }
+    }
+
+    // swiftlint:disable line_length
+    func sk1CheckTrialOrIntroPriceEligibility(_ productIdentifiers: [String],
+                                              completionBlock receiveEligibility: @escaping ReceiveIntroEligibilityBlock) {
+        // swiftlint:enable line_length
+        receiptFetcher.receiptData(refreshPolicy: .onlyIfEmpty) { maybeData in
+            if #available(iOS 12.0, macOS 10.14, macCatalyst 13.0, tvOS 12.0, watchOS 6.2, *),
+               let data = maybeData {
+                self.sk1ModernEligibilityHandler(maybeReceiptData: data,
+                                              productIdentifiers: productIdentifiers,
+                                              completionBlock: receiveEligibility)
+            } else {
+                self.backend.getIntroEligibility(appUserID: self.appUserID,
+                                                 receiptData: maybeData ?? Data(),
+                                                 productIdentifiers: productIdentifiers) { result, maybeError in
+                    if let error = maybeError {
+                        Logger.error(String(format: Strings.purchase.unable_to_get_intro_eligibility_for_user,
+                                            error.localizedDescription))
+                    }
+                    self.operationDispatcher.dispatchOnMainThread {
+                        receiveEligibility(result)
+                    }
+                }
+            }
+        }
+    }
+
+    // swiftlint:disable line_length
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func sk2CheckTrialOrIntroPriceEligibility(_ productIdentifiers: [String],
```

Comment:
one more note: we might wanna add a method to productsManager that _only_ fetches from SK2 (just like we have for SK1) for added safety: 
in this case, if the productsManager decided to fetch with sk1, [then this line would fail](https://github.com/RevenueCat/purchases-ios/pull/807/files#diff-c2f6206ed3d98e5272385745c0b4cf519a38140009ed9fb4e55dcf69cfc9aaedR90), and things won't work as expected. 
so we could have the alternative method just return `SK2ProductDetails` instances

Reply:
Made the changes. I noticed the sk1 `products(withIdentifiers)` function returns SK1 SKProducts, instead of `ProductDetails`. It should probably return ProductDetails too I think. I didn't want to make the change in this PR and create conflicts with your PR. What do you think.